modules/basicfuncs/numeric-funcs.c

```diff
@@ -56,83 +111,204 @@ tf_num_parse(gint argc, GString *argv[],
 static void
 tf_num_plus(LogMessage *msg, gint argc, GString *argv[], GString *result)
 {
-  gint64 n, m;
+  struct value n, m;
 
   if (!tf_num_parse(argc, argv, "+", &n, &m))
     {
       g_string_append_len(result, "NaN", 3);
       return;
     }
 
-  format_int64_padded(result, 0, ' ', 10, n + m);
+  if (n.value_type == VALUE_TYPE_GINT64 && m.value_type == VALUE_TYPE_GINT64)
+    {
+      format_int64_padded(result, 0, ' ', 10, value_as_int(n) + value_as_int(m));
+    }
+  else
+    {
+      gdouble res = value_as_double(n) + value_as_double(m);
+      g_string_append_printf(result, "%f", res);
+    }
 }
 
 TEMPLATE_FUNCTION_SIMPLE(tf_num_plus);
 
 static void
 tf_num_minus(LogMessage *msg, gint argc, GString *argv[], GString *result)
 {
-  gint64 n, m;
+  struct value n, m;
 
   if (!tf_num_parse(argc, argv, "-", &n, &m))
     {
       g_string_append_len(result, "NaN", 3);
       return;
     }
 
-  format_int64_padded(result, 0, ' ', 10, n - m);
+  if (n.value_type == VALUE_TYPE_GINT64 && m.value_type == VALUE_TYPE_GINT64)
+    {
+      format_int64_padded(result, 0, ' ', 10, value_as_int(n) - value_as_int(m));
+    }
+  else
+    {
+      gdouble res = value_as_double(n) - value_as_double(m);
+      g_string_append_printf(result, "%f", res);
```

Comment:
Regarding the precision, while the `double` could hold a nice range. Saving that into that as string has impact on memory usage. It would be really nice to know the usecases it could come up.

I would suggest an arbitrary number to be 20 fractal digit.
This is based on assumption that *float* is used for time (like fraction of seconds), and the current limitation of such measurement according to wikipedia is somewhere near the `10e-17` range. [Atomic clock](https://en.wikipedia.org/wiki/Atomic_clock#Secondary_representations_of_the_second)
Note: this is a really big reduction specially for `double`, but keep in mind that each new digit is stored as string
I am happy with other values, this is more like a conversation starter.

What precision would you suggest ?

Reply:
My use case is related to milliseconds.  Because of the roundings, I assumed a 6 digits precision was okayish (as you say, syslog-ng is not a scientific tool), but for sure, more precision is probably better and 20 digits seems a reasonable choice.